### PR TITLE
Add Spresense board

### PR DESCRIPTION
**Pull request by kamtom480**
Add support for Spresense. It is a compact development board based on Sony’s power-efficient multicore microcontroller CXD5602. This port uses the built and exported NuttX SDK.

Supported features:

- analogio
- board
- busio
- digitalio
- microcontroller
- os
- pulseio
- rtc
- supervisor
- time

Unsupported features:

- audioio
- audiobusio
- i2cslave
- rotaryio
- touchio
- gamepad
- neopixel_write
- nvm
- displayio
- freqencyio

Refer to [developer.sony.com/develop/spresense/](https://developer.sony.com/develop/spresense/) for further information about this board.

### CODE REVIEW

File: ports/spresense/Makefile [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	$(ECHO) "Creating $@"
+	$(SPRESENSE_SDK)/sdk/tools/linux/mkspk -c 2 $(BUILD)/circuitpython.elf nuttx $(BUILD)/circuitpython.spk
+
+flash: $(BUILD)/circuitpython.spk
+	$(ECHO) "Writing $< to the board"
+	$(SPRESENSE_SDK)/sdk/tools/flash.sh -c /dev/ttyUSB0 $(BUILD)/circuitpython.spk
+
+flash-bootloader: $(SPRESENSE_SDK) $(FIRMWARE)
+	$(ECHO) "Writing loader to the board"
+	$(SPRESENSE_SDK)/sdk/tools/flash.sh -l $(FIRMWARE) -c /dev/ttyUSB0
```
Comment (tannewt) [inline ports/spresense/Makefile]: I wasn't able to do this step on Mac. The kernel spun up for a while and then the script produced this error:

```
> spresense-exported-sdk/sdk/tools/flash.sh -l spresense-exported-sdk/firmware -c /dev/tty.usbmodemE594E79FECF32
Traceback (most recent call last):
  File "site-packages/pyserial-2.7-py3.6.egg/serial/serialposix.py", line 468, in read
OSError: [Errno 6] Device not configured

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "flash_writer.py", line 550, in <module>
  File "flash_writer.py", line 490, in main
  File "flash_writer.py", line 357, in cancel_autoboot
  File "flash_writer.py", line 286, in readline
  File "site-packages/pyserial-2.7-py3.6.egg/serial/serialposix.py", line 480, in read
TypeError: 'OSError' object is not subscriptable
[60064] Failed to execute script flash_writer
```

Any suggestions? Do I need to make sure jumpers are in a particular spot? ladyada may have used the board before me.

File: ports/spresense/common-hal/microcontroller/Pin.c [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+#include "shared-bindings/microcontroller/Pin.h"
+
+typedef struct {
+    const mcu_pin_obj_t *pin;
+    bool reset;
+    bool free;
+} pin_status_t;
+
+const mcu_pin_obj_t pin_D0 = PIN(PIN_UART2_RXD, false);
```
[the remark this answers is not in the document]
Comment (kamtom480) [reply ports/spresense/common-hal/microcontroller/Pin.c]: I changed the port name to cxd56 and added Spresense as a board. In the port I use cxd56 pin names, while in board I use Spresense pin names.

File: ports/spresense/Makefile [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+	$(ECHO) "AR $@"
+	$(Q)$(AR) rcs $(BUILD)/libmpy.a $(OBJ)
+
+$(BUILD)/circuitpython.elf: $(BUILD)/libmpy.a
+	$(ECHO) "LD $@"
+	$(Q)$(LD) $(LDFLAGS)
+
+$(BUILD)/circuitpython.spk: $(BUILD)/circuitpython.elf
+	$(ECHO) "Creating $@"
+	$(SPRESENSE_SDK)/sdk/tools/linux/mkspk -c 2 $(BUILD)/circuitpython.elf nuttx $(BUILD)/circuitpython.spk
```
[the remark this answers is not in the document]
Comment (kamtom480) [reply ports/spresense/Makefile]: I changed Makefile to work across OS. You can try now.

File: ports/spresense/Makefile (hunk 1 above)
Comment (kamtom480) [reply ports/spresense/Makefile]: 1. Have you installed the serial driver?
2. Have you connected USB to main board? USB connected to the extension board is not intended for flashing new image.
3. Try using /dev/tty.SLAB_USBtoUART as a serial device.

File: ports/spresense/common-hal/analogio/AnalogIn.c [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    read(analogin_dev[self->number].fd, &value, sizeof(value));
+
+    return value;
+}
+
+// Reference voltage is a fixed value which is depending on the board.
+// e.g.)
+// - Reference Voltage of A4 and A5 pins on Main Board is 0.7V.
+// - Reference Voltage of A0 ~ A5 pins on External Interface board
+//   is selected 3.3V or 5.0V by a IO Volt jumper pin.
```
[the remark this answers is not in the document]
Comment (kamtom480) [reply ports/spresense/common-hal/analogio/AnalogIn.c]: I didn't change it. This function returns a float variable. Do you want me to change the CircuitPython source code to interpret 0.0 here as None?

**Comment by kamtom480** [comment]
Some changes. I added Spresense to your tools, but I'm not sure if it's correct. I'm also not sure what to do with Reference Voltage.

File: ports/spresense/Makefile (hunk 1 above)
Comment (tannewt) [reply ports/spresense/Makefile]: I hadn't updated the serial driver and may have been using the wrong serial device. I got it working today though! 🎉

File: ports/spresense/common-hal/analogio/AnalogIn.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (tannewt) [reply ports/spresense/common-hal/analogio/AnalogIn.c]: Ya, I think changing shared-bindings to interpret 0.0 as None makes sense. That way an exception will be thrown if the user code uses it in any calculations.

**Comment by tannewt** [comment]
Workflow looks good and I successfully got a REPL on my board! This is ready after tweaking the analog reference voltage.

**Comment by kamtom480** [comment]
@tannewt I have added requested changes.

**Pull request merged**